dashboard/app/models/user_level.rb

```diff
@@ -130,28 +130,29 @@ def before_unsubmit
     end
   end
 
+  # `locked` is a virtual attribute because it relies on `unlocked_at` to
+  # automatically return `true` after `AUTOLOCK_PERIOD`
+  def locked
+    unlocked_at.nil? || unlocked_at < AUTOLOCK_PERIOD.ago
+  end
+
   def locked=(val)
-    new_time = if val
-                 nil
-               else
-                 Time.now
-               end
-    self.unlocked_at = new_time
-  end
-
-  def locked?
-    if unlocked_at.nil?
-      return true
-    else
-      return unlocked_at < AUTOLOCK_PERIOD.ago
-    end
+    self.unlocked_at = val ? nil : Time.now
   end
 
+  # this is the "locked" value we return to the client. if the lesson isn't
+  # lockable or we are an authorized teacher, we always return `false`.
   def show_as_locked?(stage)
     return false unless stage.lockable?
     return false if user.authorized_teacher?
-    return false if readonly_answers && !locked?
```

Comment:
this line was redundant, since if `readonly_answers=true`, this becomes `return false if !locked else return locked`, which is the same as `return locked`.

Reply:
So glad we're resulting in a state that doesn't consider submit or readonly in this method...